### PR TITLE
Move more windows to UI project

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
Remaining:
- How to pass callbacks/function pointers
- Rewrite Intent implementation
- Parse remaining window intents

File: src/openrct2/editor.c [hunk 1]
```diff
@@ -113,8 +113,11 @@ void editor_load()
 void editor_convert_save_to_scenario()
 {
     tool_cancel();
-    window_loadsave_open(LOADSAVETYPE_LOAD | LOADSAVETYPE_GAME, NULL);
-    window_loadsave_set_loadsave_callback(editor_convert_save_to_scenario_callback);
+    Intent * intent = intent_create(WC_LOADSAVE);
+    intent_set_uint(intent, INTENT_EXTRA_LOADSAVE_TYPE, LOADSAVETYPE_LOAD | LOADSAVETYPE_GAME);
+    intent_set_pointer(intent, INTENT_EXTRA_CALLBACK, editor_convert_save_to_scenario_callback);
```
Comment (janisozaur) [inline src/openrct2/editor.c]: AppVeyor:

> editor.c(118): warning C4152: nonstandard extension, function/data pointer conversion in expression [C:\projects\openrct2-ject9\src\openrct2\libopenrct2.vcxproj]

File: src/openrct2/game.c [hunk 2]
```diff
@@ -1397,8 +1412,11 @@ void game_load_or_quit_no_save_prompt()
         if (gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) {
             load_landscape();
         } else {
-            window_loadsave_open(LOADSAVETYPE_LOAD | LOADSAVETYPE_GAME, NULL);
-            window_loadsave_set_loadsave_callback(game_load_or_quit_no_save_prompt_callback);
+            Intent * intent = intent_create(WC_LOADSAVE);
+            intent_set_uint(intent, INTENT_EXTRA_LOADSAVE_TYPE, LOADSAVETYPE_LOAD | LOADSAVETYPE_GAME);
+            intent_set_pointer(intent, INTENT_EXTRA_CALLBACK, game_load_or_quit_no_save_prompt_callback);
```
Comment (janisozaur) [inline src/openrct2/game.c]: AppVeyor:

> game.c(1417): warning C4152: nonstandard extension, function/data pointer conversion in expression [C:\projects\openrct2-ject9\src\openrct2\libopenrct2.vcxproj]

**Comment by marijnvdwerf** [comment]
I have no idea what's the proper way to store the callbacks

**Comment by IntelOrca** [comment]
You need to explicitly cast to `void *`.

**Comment by marijnvdwerf** [comment]
@janisozaur @IntelOrca Can you go over this PR again? I think everything should be fine now, except for the Intent needing re-implementation

File: src/openrct2-ui/WindowManager.cpp [hunk 3]
```diff
@@ -93,6 +103,84 @@ class WindowManager final : public IWindowManager
...
+        case WC_FIRE_PROMPT:
+            return window_staff_fire_prompt_open((rct_peep*)intent->GetPointerExtra(INTENT_EXTRA_PEEP));
+        case WC_INSTALL_TRACK:
+            return window_install_track_open(intent->GetStringExtra(INTENT_EXTRA_PATH));
+        case WC_GUEST_LIST:
+            return window_guest_list_open_with_filter(intent->GetSIntExtra(INTENT_EXTRA_GUEST_LIST_FILTER), intent->GetSIntExtra(INTENT_EXTRA_RIDE_ID));
+        case WC_LOADSAVE:
+        {
+            uint32 type = intent->GetUIntExtra(INTENT_EXTRA_LOADSAVE_TYPE);
+            utf8 *defaultName = intent->GetStringExtra(INTENT_EXTRA_PATH)
```
Comment (janisozaur) [inline src/openrct2-ui/WindowManager.cpp]: `const`, space after `*`

**Comment by janisozaur** [comment]
Why there are both: `intent_set_uint` and `intent.putExtra`?

Needs some `const`s here and there, but other than that LGTM

**Comment by marijnvdwerf** [comment]
@janisozaur intents are created both by C code, as well as C++ code.

**Comment by marijnvdwerf** [comment]
@IntelOrca I attempted to refactor the intents. Was this what you were thinking of?

**Comment by IntelOrca** [comment]
@marijnvdwerf yes, but also to assert the correct type when fetching the values. So if you get a pointer, but the data type is string, throw assert.

**Comment by marijnvdwerf** [comment]
Moved dialogs:
- Banner.cpp
- DemolishRidePrompt.cpp
- Error.cpp
- Finances.cpp
- GuestList.cpp
- InstallTrack.cpp
- LoadSave.cpp
- NewCampaign.cpp
- Park.cpp
- Player.cpp
- Sign.cpp
- StaffFirePrompt.cpp
- TitleCommandEditor.cpp
- TitleEditor.cpp
- TitleScenarioSelect.cpp
- TrackDesignManage.cpp
- TrackDesignPlace.cpp

**Comment by PFCKrutonium** [comment]
I went through all the moved Windows (as far as I know anyway) and everything seems to be working correctly. 👍

**Comment by marijnvdwerf** [comment]
Is there anything left that still need to be attended to?

File: src/openrct2/windows/Intent.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
+#ifndef OPENRCT2_INTENT_H
+#define OPENRCT2_INTENT_H
+
+#include "../common.h"
+#ifdef __cplusplus
+extern "C" {
```
Comment (janisozaur) [inline src/openrct2/windows/Intent.h]: This should not be used in headers anymore

Comment (janisozaur) [reply src/openrct2/windows/Intent.h]: For including other headers, I mean

File: src/openrct2/windows/Intent.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    Intent * putExtra(uint32 key, sint32 value);
+    Intent * putExtra(uint32 key, utf8string value);
+};
+
+
+extern "C" {
+#endif // __cplusplus
+
+#ifndef __cplusplus
+    typedef int Intent;
```
Comment (janisozaur) [inline src/openrct2/windows/Intent.h]: Typedefs are not concerned by extern "C", so you can move it out of here

File: src/openrct2/windows/ObjectLoadError.cpp [hunk 6]
```diff
@@ -95,9 +95,9 @@ static rct_window_event_list window_object_load_error_events = {
     window_object_load_error_scrollpaint
 };
 
-rct_object_entry * invalid_entries = nullptr;
-sint32 highlighted_index = -1;
-utf8* file_path = nullptr;
+static rct_object_entry * invalid_entries = nullptr;
```
Comment (janisozaur) [inline src/openrct2/windows/ObjectLoadError.cpp]: Names of statics should start with an underscore

**Comment by marijnvdwerf** [comment]
@janisozaur Can you check the Intent header again?

File: src/openrct2/windows/Intent.h [hunk 7]
```diff
@@ -42,15 +36,15 @@ class Intent
     Intent * putExtra(uint32 key, sint32 value);
     Intent * putExtra(uint32 key, utf8string value);
 };
+#else
+typedef void Intent;
```
Comment (janisozaur) [inline src/openrct2/windows/Intent.h]: Oh, I only just noticed… you have a class `Intent` above that you want to use as opaque pointer in C code? Such case could use a comment explaining things.

I'll need to check if that's the best/safe way to do this.

**Comment by marijnvdwerf** [comment]
@janisozaur Is this solution better?

**Comment by janisozaur** [comment]
Clicking "Load game" button in main menu leads to exception being thrown and unhandled, leading to `std::terminate` with this backtrace:

```
#0  0x00007ffff4ea58a0 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff4ea6f09 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff5a37d77 in __gnu_cxx::__verbose_terminate_handler () at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007ffff5a358e6 in __cxxabiv1::__terminate (handler=<optimized out>) at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007ffff5a35933 in std::terminate () at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00007ffff5a35b88 in __cxxabiv1::__cxa_throw (obj=obj@entry=0x5555580af950, tinfo=0x7ffff5d230f8 <typeinfo for std::out_of_range>, dest=0x7ffff5a4b640 <std::out_of_range::~out_of_range()>) at /build/gcc-multilib/src/gcc/libstdc++-v3/libsupc++/eh_throw.cc:93
#6  0x00007ffff5a60223 in std::__throw_out_of_range (__s=0x555555f19717 "map::at") at /build/gcc-multilib/src/gcc/libstdc++-v3/src/c++11/functexcept.cc:82
#7  0x000055555595e9d5 in std::map<unsigned int, IntentData, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, IntentData> > >::at (this=0x5555580af028, __k=@0x7fffffffc644: 2) at /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/bits/stl_map.h:533
#8  0x000055555595e663 in Intent::GetStringExtra (this=0x5555580af020, key=2) at ../src/openrct2/windows/Intent.cpp:81
#9  0x00005555556f08d3 in WindowManager::OpenIntent (this=0x5555568f5810, intent=0x5555580af020) at ../src/openrct2-ui/WindowManager.cpp:171
#10 0x00005555558661a6 in context_open_intent (intent=0x5555580af020) at ../src/openrct2/Context.cpp:1016
#11 0x000055555575d972 in game_load_or_quit_no_save_prompt () at ../src/openrct2/game.c:1418
#12 0x00005555557387a0 in window_save_prompt_open () at ../src/openrct2-ui/windows/SavePrompt.cpp:129
#13 0x00005555556f053d in WindowManager::OpenWindow (this=0x5555568f5810, wc=14 '\016') at ../src/openrct2-ui/WindowManager.cpp:82
#14 0x0000555555866090 in context_open_window (wc=14 '\016') at ../src/openrct2/Context.cpp:998
#15 0x000055555575db7b in game_load_or_quit (eax=0x7fffffffc954, ebx=0x7fffffffc950, ecx=0x7fffffffc94c, edx=0x7fffffffc948, esi=0x7fffffffc944, edi=0x7fffffffc940, ebp=0x7fffffffc970) at ../src/openrct2/game.c:902
#16 0x000055555575b1b9 in game_do_command_p (command=5, eax=0x7fffffffc954, ebx=0x7fffffffc950, ecx=0x7fffffffc94c, edx=0x7fffffffc948, esi=0x7fffffffc944, edi=0x7fffffffc940, ebp=0x7fffffffc970) at ../src/openrct2/game.c:575
#17 0x000055555575acfd in game_do_command (eax=0, ebx=1, ecx=0, edx=0, esi=5, edi=0, ebp=0) at ../src/openrct2/game.c:480
#18 0x0000555555745580 in window_title_menu_mouseup (w=0x55555659e100 <g_window_list+1248>, widgetIndex=1) at ../src/openrct2-ui/windows/TitleMenu.cpp:161
#19 0x0000555555777d1a in window_event_mouse_up_call (w=0x55555659e100 <g_window_list+1248>, widgetIndex=1) at ../src/openrct2/interface/window.c:1924
#20 0x000055555575ef35 in input_state_widget_pressed (x=842, y=881, state=2, widgetIndex=1, w=0x55555659e100 <g_window_list+1248>, widget=0x5555561fadd0 <window_title_menu_widgets+32>) at ../src/openrct2/input.c:1277
#21 0x000055555575e15a in game_handle_input_mouse (x=842, y=881, state=2) at ../src/openrct2/input.c:321
#22 0x000055555575dd25 in game_handle_input () at ../src/openrct2/input.c:128
#23 0x0000555555942c6c in TitleScreen::Update (this=0x555558db8d40) at ../src/openrct2/title/TitleScreen.cpp:148
#24 0x000055555586a961 in OpenRCT2::Context::Update (this=0x555556af7a40) at ../src/openrct2/Context.cpp:686
#25 0x000055555586a84c in OpenRCT2::Context::RunFixedFrame (this=0x555556af7a40) at ../src/openrct2/Context.cpp:615
#26 0x000055555586a65f in OpenRCT2::Context::RunFrame (this=0x555556af7a40) at ../src/openrct2/Context.cpp:588
#27 0x0000555555869d2b in OpenRCT2::Context::RunGameLoop (this=0x555556af7a40) at ../src/openrct2/Context.cpp:559
#28 0x000055555586990f in OpenRCT2::Context::Launch (this=0x555556af7a40) at ../src/openrct2/Context.cpp:535
#29 0x00005555558676a9 in OpenRCT2::Context::RunOpenRCT2 (this=0x555556af7a40, argc=1, argv=0x7fffffffcf18) at ../src/openrct2/Context.cpp:157
#30 0x00005555556e521c in main (argc=1, argv=0x7fffffffcf18) at ../src/openrct2-ui/Ui.cpp:60
```

I have not investigated yet whether that's due to some technical issue or simply an unhandled case.

**Comment by marijnvdwerf** [comment]
Ah, unhandled case! Already on it

**Comment by marijnvdwerf** [comment]
@janisozaur Feel free to retry

**Pull request merged**